csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -197,10 +316,63 @@ public CompressionResult CompressPayload()
             _compressionMinSize = compressionMinSize;
         }
 
-        internal void Finish(OutputStream.Position encapsulationEnd)
+        private OutputStream WriteBinaryContextEntryKey(int key)
         {
-            Size = Payload.GetByteCount();
+            if (Protocol == Protocol.Ice1)
+            {
+                throw new NotSupportedException("binary context is not supported with Ice1 protocol");
+            }
+
+            if (IsSealed)
+            {
+                throw new InvalidOperationException("cannot modify a sealed frame");
+            }
+
+            OutputStream.Position encapsulationEnd = _encapsulationEnd ??
+                throw new InvalidOperationException(
+                    "binary context cannot be written before finishing the encapsulation");
+
+            if (_binaryContextKeys.Contains(key))
+            {
+                throw new ArgumentException($"key `{key}' is already in use", nameof(key));
+            }
+
+            var ostr = new OutputStream(Encoding, Payload, _binaryContextEnd ?? encapsulationEnd);
+            if (_binaryContextKeys.Count == 0)
+            {
+                ostr.StartFixedLengthSize(2); // 2-bytes size place holder
```

Comment:
As commented earlier, I don't like this optimization. It's also weird to write the size of the dictionary inside this `WriteBinaryContextEntryKey` helper method.